File: webapp/lib/tts/gemini-tts.ts
import "server-only";
import { getProviderKey } from "@/lib/secrets";
import { GEMINI_TTS_DEFAULT_MODEL } from "@/lib/tts/gemini-voices";

// Port of TTSCore/Sources/TTSCore/GeminiTTS.swift — multi-speaker conversation
// synthesis via Gemini's generateContent API.

export { GEMINI_TTS_VOICES, GEMINI_TTS_DEFAULT_MODEL } from "@/lib/tts/gemini-voices";

export type ConversationSpeaker = "speaker1" | "speaker2";

const DEFAULT_SPEAKER_LABEL: Record<ConversationSpeaker, string> = {
  speaker1: "Speaker 1",
  speaker2: "Speaker 2",
};

export type ConversationDialogueLine = {
  speaker: ConversationSpeaker;
  text: string;
};

export class GeminiTTSError extends Error {}

/**
 * Labels used in the Gemini transcript + multiSpeakerVoiceConfig.speaker fields.
 * Prefer character names (Kaito/Mika) — generic "Speaker 1/2" is more prone to the
 * known multi-speaker voice-swap quirk on the first turn.
 */
function resolveSpeakerLabels(params: {
  speaker1Name?: string | null;
  speaker2Name?: string | null;
}): Record<ConversationSpeaker, string> {
  const raw1 = params.speaker1Name?.trim() || "";
  const raw2 = params.speaker2Name?.trim() || "";
  if (raw1 && raw2 && raw1.toLowerCase() !== raw2.toLowerCase()) {
    return { speaker1: raw1, speaker2: raw2 };
  }
  return { ...DEFAULT_SPEAKER_LABEL };
}

function buildTranscript(
  lines: ConversationDialogueLine[],
  labels: Record<ConversationSpeaker, string>
): string {
  const body = lines
    .map((line) => `${labels[line.speaker]}: ${line.text.trim()}`)
    .join("\n");
  // Brief style cue helps Gemini keep voices attached to the right speaker.
  const cue = `TTS dialogue. ${labels.speaker1} uses voice A; ${labels.speaker2} uses voice B. Keep each speaker's voice consistent for every line.\n\n`;
  return `${cue}## Transcript:\n${body}`;
}

/** Strips a 44-byte WAV header if the response happens to be WAV-wrapped. */
function stripWavHeaderIfPresent(data: Buffer): Buffer {
  if (data.length > 44 && data.toString("ascii", 0, 4) === "RIFF") {
    return data.subarray(44);
  }
  return data;
}

export async function synthesizeGeminiConversation(params: {
  lines: ConversationDialogueLine[];
  speaker1Voice: string;
  speaker2Voice: string;
  /** Display names used as Gemini speaker aliases (e.g. Kaito / Mika). */
  speaker1Name?: string | null;
  speaker2Name?: string | null;
  model?: string;
  temperature?: number;
}): Promise<Buffer> {
  const apiKey = getProviderKey("gemini");
  if (!apiKey) {
    throw new GeminiTTSError("Missing Gemini API key.");
  }

  const trimmedLines = params.lines.filter((l) => l.text.trim().length > 0);
  if (trimmedLines.length === 0) {
    throw new GeminiTTSError("Add at least one dialogue line.");
  }

  const model = params.model ?? GEMINI_TTS_DEFAULT_MODEL;
  const labels = resolveSpeakerLabels({
    speaker1Name: params.speaker1Name,
    speaker2Name: params.speaker2Name,
  });
  const transcript = buildTranscript(trimmedLines, labels);

  const payload = {
    contents: [
      {
        role: "user",
        parts: [{ text: transcript }],
      },
    ],
    generationConfig: {
      temperature: params.temperature ?? 1,
      responseModalities: ["AUDIO"],
      speechConfig: {
        multiSpeakerVoiceConfig: {
          speakerVoiceConfigs: [
            {
              speaker: labels.speaker1,
              voiceConfig: {
                prebuiltVoiceConfig: { voiceName: params.speaker1Voice },
              },
            },
            {
              speaker: labels.speaker2,
              voiceConfig: {
                prebuiltVoiceConfig: { voiceName: params.speaker2Voice },
              },
            },
          ],
        },
      },
    },
  };

  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent`,
    {
      method: "POST",
      headers: {
        "x-goog-api-key": apiKey,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    }
  );

  if (!response.ok) {
    const body = await response.text().catch(() => "");
    throw new GeminiTTSError(
      `Gemini request failed (HTTP ${response.status}): ${body}`
    );
  }

  const json = await response.json();
  const parts = json?.candidates?.[0]?.content?.parts;
  if (!Array.isArray(parts)) {
    throw new GeminiTTSError("Gemini returned no audio data.");
  }

  for (const part of parts) {
    const base64 = part?.inlineData?.data;
    if (typeof base64 === "string" && base64.length > 0) {
      const audioData = Buffer.from(base64, "base64");
      return stripWavHeaderIfPresent(audioData);
    }
  }

  throw new GeminiTTSError("Gemini returned no audio data.");
}

export const GEMINI_TTS_SAMPLE_RATE = 24_000;

File: TTSCore/Sources/TTSCore/GeminiTTS.swift
//
//  GeminiTTS.swift
//  TTSCore
//
//  Multi-speaker conversation synthesis via Gemini 3.1 Flash TTS Preview.
//

import Foundation

public enum GeminiTTSVoice: String, CaseIterable, Codable, Sendable {
    case zephyr = "Zephyr"
    case puck = "Puck"
    case charon = "Charon"
    case kore = "Kore"
    case fenrir = "Fenrir"
    case leda = "Leda"
    case orus = "Orus"
    case aoede = "Aoede"
    case callirrhoe = "Callirrhoe"
    case autonoe = "Autonoe"
    case enceladus = "Enceladus"
    case iapetus = "Iapetus"
    case umbriel = "Umbriel"
    case algieba = "Algieba"
    case despina = "Despina"
    case erinome = "Erinome"
    case algenib = "Algenib"
    case rasalgethi = "Rasalgethi"
    case laomedeia = "Laomedeia"
    case achernar = "Achernar"
    case alnilam = "Alnilam"
    case schedar = "Schedar"
    case gacrux = "Gacrux"
    case pulcherrima = "Pulcherrima"
    case achird = "Achird"
    case zubenelgenubi = "Zubenelgenubi"
    case vindemiatrix = "Vindemiatrix"
    case sadachbia = "Sadachbia"
    case sadaltager = "Sadaltager"
    case sulafat = "Sulafat"
}

public enum ConversationSpeaker: String, CaseIterable, Codable, Sendable, Identifiable {
    case speaker1 = "Speaker 1"
    case speaker2 = "Speaker 2"

    public var id: String { rawValue }
}

public struct ConversationDialogueLine: Identifiable, Codable, Equatable, Sendable {
    public var id: UUID
    public var speaker: ConversationSpeaker
    public var text: String

    public init(id: UUID = UUID(), speaker: ConversationSpeaker, text: String) {
        self.id = id
        self.speaker = speaker
        self.text = text
    }
}

public enum GeminiTTSError: Error, LocalizedError {
    case missingAPIKey
    case noDialogueLines
    case invalidResponse(status: Int, body: String?)
    case noAudioInResponse
    case couldNotDecodeAudio

    public var errorDescription: String? {
        switch self {
        case .missingAPIKey:
            return "Missing Gemini API key."
        case .noDialogueLines:
            return "Add at least one dialogue line."
        case let .invalidResponse(status, body):
            if let body, !body.isEmpty {
                return "Gemini request failed (HTTP \(status)): \(body)"
            }
            return "Gemini request failed (HTTP \(status))."
        case .noAudioInResponse:
            return "Gemini returned no audio data."
        case .couldNotDecodeAudio:
            return "Could not decode Gemini audio response."
        }
    }
}

public enum GeminiConversationTTSClient {
    public static let defaultModel = "gemini-3.1-flash-tts-preview"

    public static func buildTranscript(from lines: [ConversationDialogueLine]) -> String {
        let body = lines
            .map { "\($0.speaker.rawValue): \($0.text.trimmingCharacters(in: .whitespacesAndNewlines))" }
            .joined(separator: "\n")
        return "## Transcript:\n\(body)"
    }

    public static func generatePCM(
        apiKey: String,
        lines: [ConversationDialogueLine],
        speaker1Voice: GeminiTTSVoice,
        speaker2Voice: GeminiTTSVoice,
        model: String = defaultModel,
        temperature: Double = 1
    ) async throws -> Data {
        guard !apiKey.isEmpty else { throw GeminiTTSError.missingAPIKey }

        let trimmedLines = lines.filter {
            !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
        guard !trimmedLines.isEmpty else { throw GeminiTTSError.noDialogueLines }

        let transcript = buildTranscript(from: trimmedLines)
        let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent")!

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 120

        let payload: [String: Any] = [
            "contents": [
                [
                    "role": "user",
                    "parts": [["text": transcript]],
                ],
            ],
            "generationConfig": [
                "temperature": temperature,
                "responseModalities": ["AUDIO"],
                "speechConfig": [
                    "multiSpeakerVoiceConfig": [
                        "speakerVoiceConfigs": [
                            [
                                "speaker": ConversationSpeaker.speaker1.rawValue,
                                "voiceConfig": [
                                    "prebuiltVoiceConfig": ["voiceName": speaker1Voice.rawValue],
                                ],
                            ],
                            [
                                "speaker": ConversationSpeaker.speaker2.rawValue,
                                "voiceConfig": [
                                    "prebuiltVoiceConfig": ["voiceName": speaker2Voice.rawValue],
                                ],
                            ],
                        ],
                    ],
                ],
            ],
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw GeminiTTSError.invalidResponse(status: -1, body: nil)
        }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8)
            throw GeminiTTSError.invalidResponse(status: http.statusCode, body: body)
        }

        return try extractPCM(from: data)
    }

    private static func extractPCM(from responseData: Data) throws -> Data {
        guard
            let json = try JSONSerialization.jsonObject(with: responseData) as? [String: Any],
            let candidates = json["candidates"] as? [[String: Any]],
            let first = candidates.first,
            let content = first["content"] as? [String: Any],
            let parts = content["parts"] as? [[String: Any]]
        else {
            throw GeminiTTSError.noAudioInResponse
        }

        for part in parts {
            guard
                let inlineData = part["inlineData"] as? [String: Any],
                let base64 = inlineData["data"] as? String,
                let audioData = Data(base64Encoded: base64)
            else { continue }

            let mimeType = inlineData["mimeType"] as? String ?? "audio/L16;rate=24000"
            if mimeType.lowercased().contains("wav") || audioData.starts(with: [0x52, 0x49, 0x46, 0x46]) {
                return stripWAVHeader(from: audioData) ?? audioData
            }
            return audioData
        }

        throw GeminiTTSError.noAudioInResponse
    }

    private static func stripWAVHeader(from data: Data) -> Data? {
        guard data.count > 44 else { return nil }
        return data.subdata(in: 44..<data.count)
    }
}
